### PR TITLE
feat: Refactor to make it easy to extend new kinds

### PR DESCRIPTION
**Pull request by gaocegege**
Signed-off-by: Ce Gao <gaoce@caicloud.io>

**What this PR does / why we need it**:

This PR introduces a new Interface Provider, to make it easy to add new kinds.

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/865)

### CODE REVIEW

File: pkg/job/v1alpha3/kubeflow/kubeflow.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+
+	"github.com/kubeflow/katib/pkg/controller.v1alpha3/consts"
+)
+
+const (
+	PyTorchJobWorkerContainerName = "pytorch"
+
+	TFJobWorkerContainerName = "tensorflow"
```
Comment (johnugeorge) [inline pkg/job/v1alpha3/kubeflow/kubeflow.go]: Should we reference this from tf operator repo and PyTorch operator repo?

Comment (gaocegege) [reply pkg/job/v1alpha3/kubeflow/kubeflow.go]: SGTM. I think so.

**Comment by gaocegege** [comment]
Addressed the comment, PTAL @johnugeorge

File: pkg/job/v1alpha3/consts.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	TFJobRoleLabel = "tf-job-role"
+	// PyTorchJobRoleLabel is deprecated in kubeflow 0.7, but we need to be compatible.
+	PyTorchJobRoleLabel = "pytorch-job-role"
+)
+
+// JobRoleMap is the map which is used to determin if the replica is master.
+// Katib will inject metrics collector into master replica.
+var JobRoleMap = map[string][]string{
+	// Job kind does not support distributed training, thus no master.
+	consts.JobKindJob:     {},
```
Comment (johnugeorge) [inline pkg/job/v1alpha3/consts.go]: @gaocegege Just wondering why don't we keep all job related constants in this file?

Comment (gaocegege) [reply pkg/job/v1alpha3/consts.go]: SGTM

Comment (gaocegege) [reply pkg/job/v1alpha3/consts.go]: Does it work for you? @johnugeorge

Comment (johnugeorge) [reply pkg/job/v1alpha3/consts.go]: I feel that we have defined too many constants :)

**Comment by johnugeorge** [comment]
Ready to be merged?

**Comment by gaocegege** [comment]
I think so.

**Pull request merged**